### ISSUE TITLE
Consider switching to single artefact dir by default for improved typings.

### ISSUE THREAD

**Issue by alloy** [facebook/relay#2509]
Currently the default is that artefact directories are colocated with the modules from which the GraphQL document was extracted. However, for people that do not use the Haste module system (aka most people outside FB) this means that typings for fragment references won’t work (or rather, they are typed as `any`).

With the introduction of the language plugin infrastructure, the option to have a single artefact directory was introduced, specifically to solve this issue in a simple and stable way. i.e. artefacts can always assume that other artefacts are sibling files and thus can always use relative imports.

Seeing as this option benefits _most_ people, it might make sense to consider making that the default.

@jstejada @kassens Interested to hear your thoughts on this.

**Comment by AndrewIngram** [comment]
This benefits flow users too, because with a single artefact dir it's possible to configure flow to read all the definitions (which would need to be changed to use `declare`), this would allow referencing the generated types without explicitly importing them.

**Comment by AugustinLF** [comment]
I'm a flow user. The main issue with the single artifact dir is that it makes it harder to import the types from the corresponding component (I really do not want to use relative imports for that). But I guess that could be solved by using the module resolver feature. That should work. it breaks (or make complicated) stuff like autocomplete when importing, but I guess it's an acceptable drawback. @AndrewIngram's solution could also solve this issue!

The other problem I see is that when committing the files, I'm struggling to make my team checks the diff (I found that looking at the types is pretty interesting, since it helps to find schema mistakes, mostly around nullable types), and I think with this new system, it'll be even less likely, since the generated file will be far from the modified file.

**Comment by AndrewIngram** [comment]
That's an interesting take on checking the diff. We've gone the other way and have Github hide the changes in generated files by default. I occasionally look at the generated files during development, but almost never during code review.

**Comment by alloy** [comment]
@AugustinLF Good point re the import path 👍 Indeed, we opted to use a module resolver. Auto completion still works fine for us, but not sure if that’s because TS and vscode integrate well automatically and it’s harder with Flow. I would assume that it should be very doable with other combinations and it’s mostly a matter of us pointing to the right docs on it or maybe we can even include some suggestions in the Relay docs?

**Comment by alloy** [comment]
As for checking the diff, we do exactly the same thing as @AndrewIngram.

In your situation, maybe naming the single artefact dir in such a way that it’s always at the top of your change-set could help people seeing those diffs?

**Comment by mrtnzlml** [comment]
Hi @alloy - what do you do in case of 2 fragments with the same name (but in different directories)? I think it's quite common in a large codebase. Relay throws `Duplicate definition of ' ... '.` in this case and you have to change the fragment name (?). Collocation works well because of this convention but it starts to be more complicated when you try to add everything generated into one file.

**Comment by alloy** [comment]
I was under the impression that Relay already required unique names, like that error you showed. Is that not the case?

**Comment by alloy** [comment]
(Btw it doesn’t all go in one file, rather in a single directory)

**Comment by mrtnzlml** [comment]
Yes, you are right. I somehow missed this. 🙈 So it's not that common after all... 😄

**Comment by alloy** [comment]
Haha, always good to check! So thanks for the feedback 👍
